File: ios/StakDemo/Session.swift
import SwiftUI

/// Persisted sign-in state (user, 2026-08-23): a user who already signed
/// in is not asked to sign in again - the splash goes straight to Home;
/// a first-time user is taken to create an account. Backed by
/// UserDefaults until the real auth backend lands (then this is the
/// seam that holds the token + the profile it returns). Mirrors android
/// ui/Session.kt.
final class Session: ObservableObject {
	static let shared = Session()

	private static let keySignedIn = "stak.signedIn"
	private static let keyName = "stak.displayName"
	private static let keyRisk = "stak.riskStyle"
	/// Legacy: earlier builds kept the photo bytes here; read once and moved
	/// to the photo file (see loadPhoto).
	private static let keyPhoto = "stak.photoData"
	private static let keyDemo = "stak.demoAccount"
	private static let keyPicks = "stak.brandPicks"
	private static let keyGoal = "stak.goalAnswer"
	private static let keyRiskAnswer = "stak.riskAnswer"

	@Published private(set) var signedIn: Bool

	/// True when this launch started already signed in - the returning-user path.
	private(set) var resumedSignedIn: Bool
	/// Which account this is (product audit, 2026-09-05; mirrors Android): Sign in = the
	/// DEMO account with the authored history, Create account = a NEW account that starts
	/// empty and earns its numbers. Persisted with the sign-in.
	@Published private(set) var demoAccount: Bool

	private init() {
		let d = UserDefaults.standard
		// A @Published property can be assigned but not READ through self
		// until every stored property is initialized - both flags come
		// from the local instead.
		let wasSignedIn = d.bool(forKey: Self.keySignedIn)
		signedIn = wasSignedIn
		resumedSignedIn = wasSignedIn
		demoAccount = d.object(forKey: Self.keyDemo) as? Bool ?? true
		UserProfile.shared.displayName = d.string(forKey: Self.keyName) ?? ""
		if let risk = d.string(forKey: Self.keyRisk) { UserProfile.shared.riskStyle = risk }
		UserProfile.shared.brandPicks = Set(d.stringArray(forKey: Self.keyPicks) ?? [])
		UserProfile.shared.goal = d.object(forKey: Self.keyGoal) as? Int ?? -1
		UserProfile.shared.risk = d.object(forKey: Self.keyRiskAnswer) as? Int ?? -1
		UserProfile.shared.photoData = Self.loadPhoto()
		applyAccount()
	}

	/// Seeds (demo) or clears (new account) every user-data singleton for the current account.
	func applyAccount() {
		MyStakHoldings.shared.reset(demo: demoAccount)
		PaperPortfolio.shared.reset(demo: demoAccount)
	}

	/// Sign-in CTA or account creation (09 Proceed) - remembered across launches.
	/// `demo` = the authored demo account (Sign in); false = a fresh account (Create account).
	func signIn(demo: Bool) {
		signedIn = true
		demoAccount = demo
		persist()
		applyAccount()
	}

	/// Profile edits after sign-in stay with the session.
	func saveProfile() { persist() }

	/// Log out: forget the session and the profile; next launch asks to sign in.
	func signOut() {
		signedIn = false
		resumedSignedIn = false
		demoAccount = true
		UserProfile.shared.displayName = ""
		UserProfile.shared.photoData = nil
		UserProfile.shared.riskStyle = "Growth-Oriented"
		UserProfile.shared.brandPicks = []
		UserProfile.shared.goal = -1
		UserProfile.shared.risk = -1
		let d = UserDefaults.standard
		d.removeObject(forKey: Self.keySignedIn)
		d.removeObject(forKey: Self.keyName)
		d.removeObject(forKey: Self.keyRisk)
		d.removeObject(forKey: Self.keyPhoto)
		d.removeObject(forKey: Self.keyDemo)
		d.removeObject(forKey: Self.keyPicks)
		d.removeObject(forKey: Self.keyGoal)
		d.removeObject(forKey: Self.keyRiskAnswer)
		Self.savePhoto(nil)
		applyAccount()
	}

	private func persist() {
		let d = UserDefaults.standard
		d.set(signedIn, forKey: Self.keySignedIn)
		d.set(demoAccount, forKey: Self.keyDemo)
		d.set(UserProfile.shared.displayName, forKey: Self.keyName)
		d.set(UserProfile.shared.riskStyle, forKey: Self.keyRisk)
		d.set(Array(UserProfile.shared.brandPicks).sorted(), forKey: Self.keyPicks)
		d.set(UserProfile.shared.goal, forKey: Self.keyGoal)
		d.set(UserProfile.shared.risk, forKey: Self.keyRiskAnswer)
		Self.savePhoto(UserProfile.shared.photoData)
	}

	// MARK: - Profile photo file

	/// The (downsampled, tens-of-KB) profile photo lives as a FILE in
	/// Application Support, not as a UserDefaults blob (audit 2026-09-04):
	/// defaults are for small values - a multi-MB blob there is flagged at
	/// 4 MB and re-read on every launch. Mirrors android's URI-backed
	/// UserProfile.photoUri.
	private static let photoFileName = "stak_profile_photo.jpg"

	private static var photoFileURL: URL? {
		guard let dir = try? FileManager.default.url(
			for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true
		) else { return nil }
		return dir.appendingPathComponent(photoFileName)
	}

	private static func loadPhoto() -> Data? {
		if let url = photoFileURL, let data = try? Data(contentsOf: url) { return data }
		// A blob an earlier build kept in defaults moves to the file once -
		// shrunk the way the picker now shrinks a pick, since that blob may
		// be a full-size original.
		let d = UserDefaults.standard
		guard let legacy = d.data(forKey: keyPhoto) else { return nil }
		d.removeObject(forKey: keyPhoto)
		let data = UIImage(data: legacy)?
			.preparingThumbnail(of: CGSize(width: 512, height: 512))?
			.jpegData(compressionQuality: 0.85) ?? legacy
		savePhoto(data)
		return data
	}

	private static func savePhoto(_ data: Data?) {
		guard let url = photoFileURL else { return }
		if let data {
			try? data.write(to: url, options: .atomic)
		} else {
			try? FileManager.default.removeItem(at: url)
		}
	}
}

File: android/app/src/main/java/com/stak/demo/ui/Session.kt
package com.stak.demo.ui

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Persisted sign-in state (user, 2026-08-23): a user who already signed
 * in is not asked to sign in again - the splash goes straight to Home;
 * a first-time user is taken to create an account. Backed by
 * SharedPreferences until the real auth backend lands (then this is
 * the seam that holds the token + the profile it returns).
 */
object Session {
	private const val PREFS = "stak_session"
	private const val KEY_SIGNED_IN = "signed_in"
	private const val KEY_NAME = "display_name"
	private const val KEY_PHOTO = "photo_uri"
	private const val KEY_RISK = "risk_style"
	private const val KEY_DEMO = "demo_account"
	private const val KEY_PICKS = "brand_picks"
	private const val KEY_GOAL = "goal_answer"
	private const val KEY_RISK_ANSWER = "risk_answer"

	private var prefs: SharedPreferences? = null

	var signedIn by mutableStateOf(false)
		private set

	/** True when this launch started already signed in - the returning-user path. */
	var resumedSignedIn = false
		private set

	/**
	 * Which account this is (product audit, 2026-09-05). Sign in = the DEMO
	 * account with the authored history (19 saved stocks, $10,240, #47);
	 * Create account = a NEW account that starts empty and earns its numbers.
	 * Persisted with the sign-in so a relaunch restores the same account.
	 */
	var demoAccount by mutableStateOf(true)
		private set

	/** Load once per process; restores the profile the user set before. */
	fun init(context: Context) {
		if (prefs != null) return
		val p = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
		prefs = p
		StakStore.init(context)
		signedIn = p.getBoolean(KEY_SIGNED_IN, false)
		resumedSignedIn = signedIn
		demoAccount = p.getBoolean(KEY_DEMO, true)
		UserProfile.displayName = p.getString(KEY_NAME, "") ?: ""
		UserProfile.photoUri = p.getString(KEY_PHOTO, null)
		UserProfile.riskStyle = p.getString(KEY_RISK, UserProfile.riskStyle) ?: UserProfile.riskStyle
		UserProfile.brandPicks = p.getString(KEY_PICKS, "")?.split(",")?.filter { it.isNotBlank() }?.toSet() ?: emptySet()
		UserProfile.goal = p.getInt(KEY_GOAL, -1)
		UserProfile.risk = p.getInt(KEY_RISK_ANSWER, -1)
		applyAccount()
	}

	/** Sign-in CTA or account creation (09 Proceed) - remembered across launches. */
	/** `demo` = the authored demo account (Sign in); false = a fresh account (Create account). */
	fun signIn(demo: Boolean) {
		signedIn = true
		demoAccount = demo
		persist()
		// A brand-new account starts from nothing; the demo account keeps
		// whatever it did last time it was signed in.
		if (!demo) StakStore.clearAccount(demo = false)
		applyAccount()
	}

	/** Seeds (demo) or clears (new account) every user-data singleton for the current account. */
	fun applyAccount() {
		MyStakHoldings.reset(demo = demoAccount)
		com.stak.demo.ui.simulate.PaperPortfolio.reset(demo = demoAccount)
		com.stak.demo.ui.discover.DeckSession.load()
		StakNotifications.load()
		com.stak.demo.ui.news.NewsSaves.load()
	}

	/** Profile edits after sign-in (name/photo) stay with the session. */
	fun saveProfile() = persist()

	/** Log out: forget the session and the profile; next launch asks to sign in. */
	fun signOut() {
		signedIn = false
		resumedSignedIn = false
		demoAccount = true
		UserProfile.displayName = ""
		UserProfile.photoUri = null
		UserProfile.riskStyle = "Growth-Oriented"
		UserProfile.brandPicks = emptySet()
		UserProfile.goal = -1
		UserProfile.risk = -1
		prefs?.edit()?.clear()?.apply()
		applyAccount()
	}

	private fun persist() {
		prefs?.edit()
			?.putBoolean(KEY_SIGNED_IN, signedIn)
			?.putBoolean(KEY_DEMO, demoAccount)
			?.putString(KEY_NAME, UserProfile.displayName)
			?.putString(KEY_PHOTO, UserProfile.photoUri)
			?.putString(KEY_RISK, UserProfile.riskStyle)
			?.putString(KEY_PICKS, UserProfile.brandPicks.joinToString(","))
			?.putInt(KEY_GOAL, UserProfile.goal)
			?.putInt(KEY_RISK_ANSWER, UserProfile.risk)
			?.apply()
	}
}
